internal/executor/executor.go

```diff
@@ -114,26 +114,25 @@ func (executor *Executor) RunBuild() {
 			}
 		}
 	}
-EXECUTION_LOOP:
+
+	// Even through we haven't run any commands yet, we bootstrap with the clean state
+	previousCommandSucceeded := true
+
 	for {
 		if currentStep.Name == executor.commandTo {
 			break
 		}
-		log.Printf("Executing %s...", currentStep.Name)
-		nextCommandName := executor.performStep(environment, currentStep)
-		if nextCommandName == "" {
-			log.Printf("%s command finished and instructed to exit!", currentStep.Name)
+
+		shouldRun := (currentStep.ExecutionBehaviour == api.Command_ON_SUCCESS && previousCommandSucceeded) ||
+			(currentStep.ExecutionBehaviour == api.Command_ON_FAILURE && !previousCommandSucceeded) ||
```

Comment:
This might not be as simple as that. Successful execution of an `ON_FAILURE` command will override `previousCommandSucceeded` which will cause the next `ON_SUCCESS` to execute which is wrong since the whole task failed.

Reply:
Indeed, see https://github.com/cirruslabs/cirrus-ci-agent/pull/32/commits/491a313d7d0e6d54596ac31c3eea830b8f93f39a.

Now the only potentially problematic thing I see here is that it will early exit and won't check all tasks:

https://github.com/cirruslabs/cirrus-ci-agent/blob/491a313d7d0e6d54596ac31c3eea830b8f93f39a/internal/executor/executor.go#L128-L130

Should I change this behavior too?

Reply:
Yeah, I think it should be removed. Good call!

Reply:
Changed to `continue`, see https://github.com/cirruslabs/cirrus-ci-agent/pull/32/commits/cca19d721297f3874511eb936a56e4857c69551d.